core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +668,934 @@ impl CheckedAdd for Fixed64 {
+
+	// A sensible value for this would be half of the dword size of the host machine. Since the
+	// runtime is compiled to 32bit webassembly, using 32 and 64 for single and double respectively
+	// should yield the most performance. TODO #3745 we could benchmark this and verify.
+	/// Representation of a single limb.
+	pub type Single = u32;
+	/// Representation of two limbs.
+	pub type Double = u64;
+	/// Difference in the number of bits of [`Single`] and [`Double`].
+	const SHIFT: usize = 32;
+	/// short form of _Base_. Analogous to the value 10 in base-10 decimal numbers.
+	const B: Double = Single::max_value() as Double + 1;
+
+	/// Splits a [`Double`] limb number into a tuple of two [`Single`] limb numbers.
+	pub fn split(a: Double) -> (Single, Single) {
+		let al = a as Single;
+		let ah = (a >> SHIFT) as Single;
+		(ah, al)
+	}
+
+	/// Assumed as a given primitive.
+	///
+	/// Multiplication of two singles, which at most yields 1 double.
+	pub fn mul_single(a: Single, b: Single) -> Double {
+		let a: Double = a.into();
+		let b: Double = b.into();
+		let r = a * b;
+		r
+	}
+
+	/// Assumed as a given primitive.
+	///
+	/// Addition of two singles, which at most takes a single limb of result and a carry,
+	/// returned as a tuple respectively.
+	pub fn add_single(a: Single, b: Single) -> (Single, Single) {
+		let a: Double = a.into();
+		let b: Double = b.into();
+		let q = a + b;
+		let (carry, r) = split(q);
+		(r, carry)
+	}
+
+	/// Assumed as a given primitive.
+	///
+	/// Division of double by a single limb. Always returns a double limb of quotient and a single
+	/// limb of remainder.
+	fn div_single(a: Double, b: Single) -> (Double, Single) {
+		let b: Double = b.into();
+		let q = a / b;
+		let r = a % b;
+		// both conversions are trivially safe.
+		(q, r as Single)
+	}
+
+	/// Simple wrapper around an infinitely large integer, represented as limbs of [`Single`].
+	#[derive(Clone, Default)]
+	pub struct BigUint {
+		/// digits (limbs) of this number (sorted as msb -> lsd).
+		pub(crate) digits: Vec<Single>,
+	}
+
+	impl BigUint {
+		/// Create a new instance with `size` limbs. This prevents any number with zero limbs to be
+		/// created.
+		///
+		/// The behavior of the type is undefined with zero limbs.
+		pub fn with_capacity(size: usize) -> Self {
+			Self { digits: vec![0; size.max(1)] }
+		}
+
+		/// Raw constructor from custom limbs. If `limbs` is empty, `Zero::zero()` implementation is
+		/// used.
+		pub fn from_limbs(limbs: &[Single]) -> Self {
+			if limbs.len() > 0 {
+				Self { digits: limbs.to_vec() }
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Number of limbs.
+		pub fn len(&self) -> usize { self.digits.len() }
+
+		/// A naive getter for limb at `index`. Note that the order is lsb -> msb.
+		///
+		/// #### Panics
+		///
+		/// This panics if index is out of range.
+		pub fn get(&self, index: usize) -> Single {
+			self.digits[self.len() - 1 - index]
+		}
```

Comment:
maybe I could be in favor of silently returning zero instead of panic but I'm also fine with that.

Reply:
Rational was to be act similar to a raw `[]` operator. I suppose with checked_get() it should be fine.